File: gcj2009QRC.cpp
#include <iostream>
#include <iomanip>
#include <string>

static const char S[] = "welcome to code jam";
static const int L = sizeof(S)-1;

static int solve() {
    std::string str;
    std::getline(std::cin,str);
    int dp[L+1] = {1};
    for (auto const& c: str) {
        for (int i=L-1; i>=0; --i) {
            if (c==S[i]) {
                dp[i+1] = (dp[i+1] + dp[i]) % 10000;
            }
        }
    }
    return dp[L];
}

int main ()
{
    int N;
    std::cin >> N;
    {
        std::string str;
        std::getline(std::cin,str);
    }
    for (int i=1; i<=N; ++i) {
        std::cout << "Case #" << i << ": "
        << std::setw(4) << std::setfill('0') << solve()
        << "\n";
    }
    return 0;
}

File: gcj2009QRC.hs
#!/usr/bin/env runghc
import Control.Monad
import Text.Printf

main = do
  c <- liftM lines getContents
  let ss = zipWith (\i r->"Case #"++show i++": "++r) [1..] $ solve c
  mapM_ putStrLn ss

solve (c:cs) = map f cs
  where
    t = "welcome to code jam"
    format::Int->String
    format = printf "%04d"
    f xs = format $ last $ g xs (1:replicate (length t) 0)
    g [] dp = dp
    g (x:xs) (dp:dps) = g xs $ dp : h dp dps t
      where
        h p [] [] = []
        h p (dp:dps) (s:ss) = n : h dp dps ss
          where
            n = if x==s then mod (p+dp) 10000 else dp
